pkg/observer/controller_test.go

```diff
@@ -387,3 +389,378 @@ func Test_dropInProgressNodeGroups(t *testing.T) {
+        metrics:        newMetrics(),
+        Options: Options{
+            Namespace:    "kube-system",
+            WaitInterval: 0,
+        },
+    }
+}
+
+func buildListed(ng *atlassianv1.NodeGroup, nodeNames ...string) *ListedNodeGroups {
+    var nodes []*corev1.Node
+    for _, n := range nodeNames {
+        nodes = append(nodes, &corev1.Node{ObjectMeta: v1.ObjectMeta{Name: n}})
+    }
+    return &ListedNodeGroups{NodeGroup: ng, List: nodes, Reason: "test"}
+}
+
+func TestPriority_AllSamePriority_CreatesAll(t *testing.T) {
+    scenario := test.BuildTestScenario(test.ScenarioOpts{Keys: []string{"a", "b"}, NodeCount: 1, PodCount: 1}).Flatten()
+    for _, ng := range scenario.Nodegroups {
+        ng.Spec.CycleSettings.Concurrency = 1
+        ng.Spec.Priority = 1
+    }
+
+    var objects []runtime.Object
+    for i := range scenario.Nodegroups {
+        objects = append(objects, scenario.Nodegroups[i])
+    }
+
+    obs := testObserver{changed: map[string]*ListedNodeGroups{
+        scenario.Nodegroups[0].Name: buildListed(scenario.Nodegroups[0], scenario.Nodes[0].Name),
+        scenario.Nodegroups[1].Name: buildListed(scenario.Nodegroups[1], scenario.Nodes[0].Name),
+    }}
+    ctrl := newPriorityControllerForTest(t, objects, scenario.Nodes, obs)
+
+    ctrl.Run()
+
+    lst, _ := generation.ListCNRs(ctrl.client, &client.ListOptions{Namespace: ctrl.Namespace})
+    assert.Equal(t, 2, len(lst.Items))
+}
+
+func TestPriority_FirstFailed_BlocksSecond(t *testing.T) {
+    scenario := test.BuildTestScenario(test.ScenarioOpts{Keys: []string{"a", "b"}, NodeCount: 1, PodCount: 1}).Flatten()
+    a := scenario.Nodegroups[0]
+    b := scenario.Nodegroups[1]
+    a.Spec.CycleSettings.Concurrency = 1
+    b.Spec.CycleSettings.Concurrency = 1
+    a.Spec.Priority = 0
+    b.Spec.Priority = 10
+
+    var objects []runtime.Object
+    objects = append(objects, a, b)
+
+    obs := testObserver{changed: map[string]*ListedNodeGroups{
+        a.Name: buildListed(a, scenario.Nodes[0].Name),
+        b.Name: buildListed(b, scenario.Nodes[0].Name),
+    }}
+    ctrl := newPriorityControllerForTest(t, objects, scenario.Nodes, obs)
+
+    // Create A
+    ctrl.Run()
+    lst1, _ := generation.ListCNRs(ctrl.client, &client.ListOptions{Namespace: ctrl.Namespace})
+    assert.Equal(t, 1, len(lst1.Items))
+
+    // Mark A as Failed
+    var cnrA atlassianv1.CycleNodeRequest
+    _ = ctrl.client.Get(context.TODO(), client.ObjectKey{Namespace: ctrl.Namespace, Name: lst1.Items[0].Name}, &cnrA)
+    cnrA.Status.Phase = atlassianv1.CycleNodeRequestFailed
+    _ = ctrl.client.Update(context.TODO(), &cnrA)
+
+    // Subsequent runs should never create B while A is Failed (treated as in-progress)
+    ctrl.Run()
+    ctrl.Run()
+    lst2, _ := generation.ListCNRs(ctrl.client, &client.ListOptions{Namespace: ctrl.Namespace})
+    assert.Equal(t, 1, len(lst2.Items))
+}
+
+func TestPriority_DefaultingNegativeAsZero_Sequential(t *testing.T) {
```

Comment:
Agree with @vincentportella from other comments on this PR that we should just allow negative numbers, but we should still default to 0 if not provided. That would make this test redundant.